File: doip_client.py
import socket
import struct
import time

# DoIP 常量
PROTOCOL_VERSION = 0x02
INVERSE_PROTOCOL_VERSION = 0xFD

PAYLOAD_TYPE_VEHICLE_IDENT_REQ = 0x0001
PAYLOAD_TYPE_VEHICLE_IDENT_RES = 0x0004
PAYLOAD_TYPE_ROUTING_ACTIVATION_REQ = 0x0005
PAYLOAD_TYPE_ROUTING_ACTIVATION_RES = 0x0006
PAYLOAD_TYPE_DIAGNOSTIC_MESSAGE = 0x8001
PAYLOAD_TYPE_DIAGNOSTIC_POS_ACK = 0x8002

# 诊断仪地址
TESTER_LOGICAL_ADDR = 0x0E80

def create_doip_header(payload_type, length):
    return struct.pack("!BBHL", PROTOCOL_VERSION, INVERSE_PROTOCOL_VERSION, payload_type, length)

def parse_header(data):
    return struct.unpack("!BBHL", data)

def main():
    print("=== DoIP 诊断仪模拟器 ===")
    
    # 1. 车辆发现 (UDP)
    print("\n[Step 1] 发送车辆发现请求 (UDP Broadcast)...")
    udp_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    udp_sock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
    udp_sock.settimeout(2.0)
    
    # 构建 Vehicle Identification Request 消息
    # Payload 为空
    header = create_doip_header(PAYLOAD_TYPE_VEHICLE_IDENT_REQ, 0)
    udp_sock.sendto(header, ('255.255.255.255', 13400))
    
    target_ip = None
    target_port = 13400
    vehicle_logical_addr = 0
    
    try:
        data, addr = udp_sock.recvfrom(1024)
        print(f"[UDP] 收到响应 来自 {addr}")
        
        ver, inv_ver, p_type, p_len = parse_header(data[:8])
        if p_type == PAYLOAD_TYPE_VEHICLE_IDENT_RES:
            payload = data[8:]
            vin = payload[:17].decode('utf-8', errors='ignore').strip('\x00')
            vehicle_logical_addr = struct.unpack("!H", payload[17:19])[0]
            print(f"       -> 发现车辆 VIN: {vin}")
            print(f"       -> 逻辑地址: 0x{vehicle_logical_addr:04X}")
            target_ip = addr[0]
            
    except socket.timeout:
        print("[!] 未发现车辆，请确保 server 正在运行。")
        # 如果是本地测试，可能广播收不到，尝试直接连接本地
        print("[*] 尝试直连 127.0.0.1 (本地回环测试)")
        target_ip = '127.0.0.1'
        vehicle_logical_addr = 0x1000 # 假设
        
    udp_sock.close()
    
    if not target_ip:
        return

    # 2. 建立 TCP 连接
    print(f"\n[Step 2] 连接车辆 TCP {target_ip}:{target_port}...")
    tcp_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        tcp_sock.connect((target_ip, target_port))
    except Exception as e:
        print(f"[!] 连接失败: {e}")
        return
        
    # 3. 路由激活
    print("\n[Step 3] 发送路由激活请求...")
    # Source Address (2) + Activation Type (1, 0x00=Default) + Reserved (4)
    ra_payload = struct.pack("!HB", TESTER_LOGICAL_ADDR, 0x00) + b'\x00'*4
    ra_header = create_doip_header(PAYLOAD_TYPE_ROUTING_ACTIVATION_REQ, len(ra_payload))
    tcp_sock.sendall(ra_header + ra_payload)
    
    # 接收激活响应
    resp_header = tcp_sock.recv(8)
    ver, inv_ver, p_type, p_len = parse_header(resp_header)
    resp_payload = tcp_sock.recv(p_len)
    
    if p_type == PAYLOAD_TYPE_ROUTING_ACTIVATION_RES:
        # Logical Address Tester (2) + Logical Address DoIP Entity (2) + Response Code (1) + Reserved (4)
        ta_res, sa_res, code = struct.unpack("!HHB", resp_payload[:5])
        if code == 0x10:
            print(f"[TCP] 路由激活成功! (Code 0x{code:02X})")
        else:
            print(f"[!] 路由激活失败: 0x{code:02X}")
            return
            
    # 4. 发送诊断请求
    print("\n[Step 4] 发送诊断请求 (ReadDataByIdentifier 0x22 F1 90)...")
    # 模拟 UDS: 22 F1 90 (读取 VIN)
    uds_req = b'\x22\xF1\x90'
    
    # DoIP Payload: Source Address (2) + Target Address (2) + User Data (N)
    diag_payload = struct.pack("!HH", TESTER_LOGICAL_ADDR, vehicle_logical_addr) + uds_req
    diag_header = create_doip_header(PAYLOAD_TYPE_DIAGNOSTIC_MESSAGE, len(diag_payload))
    tcp_sock.sendall(diag_header + diag_payload)
    
    # 5. 接收响应
    # 通常会先收到 ACK，然后是响应
    print("\n[Step 5] 等待响应...")
    
    while True:
        header_data = tcp_sock.recv(8)
        if not header_data:
            break
        
        ver, inv_ver, p_type, p_len = parse_header(header_data)
        payload_data = tcp_sock.recv(p_len)
        
        if p_type == PAYLOAD_TYPE_DIAGNOSTIC_POS_ACK:
            print(f"[TCP] 收到 DoIP ACK (Positive)")
        elif p_type == PAYLOAD_TYPE_DIAGNOSTIC_MESSAGE:
            # 解析 UDS 响应
            sa, ta = struct.unpack("!HH", payload_data[:4])
            uds_res = payload_data[4:]
            print(f"[TCP] 收到诊断响应: {uds_res.hex().upper()}")
            print(f"       -> 来自: 0x{sa:04X} 发给: 0x{ta:04X}")
            break # 演示结束

    print("\n=== 演示结束 ===")
    tcp_sock.close()

if __name__ == "__main__":
    main()

File: DoIP_Simulation/doip_client.cpp
#include <iostream>
#include <string>
#include <vector>
#include <cstring>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sys/time.h>
#include "doip_common.h"

const uint16_t TESTER_LOGICAL_ADDR = 0x0E80;

// 辅助函数：接收指定长度的数据
bool recv_fixed(int sock, void* buffer, size_t length) {
    size_t total_read = 0;
    char* ptr = (char*)buffer;
    while (total_read < length) {
        ssize_t n = recv(sock, ptr + total_read, length - total_read, 0);
        if (n <= 0) return false;
        total_read += n;
    }
    return true;
}

int main() {
    std::cout << "=== DoIP 诊断仪模拟器 (C++ Version) ===" << std::endl;

    // 1. 车辆发现 (UDP)
    std::cout << "\n[Step 1] 发送车辆发现请求 (UDP Broadcast)..." << std::endl;
    
    int udp_sock = socket(AF_INET, SOCK_DGRAM, 0);
    int broadcast = 1;
    setsockopt(udp_sock, SOL_SOCKET, SO_BROADCAST, &broadcast, sizeof(broadcast));
    
    // 设置超时
    struct timeval tv;
    tv.tv_sec = 2;
    tv.tv_usec = 0;
    setsockopt(udp_sock, SOL_SOCKET, SO_RCVTIMEO, (const char*)&tv, sizeof tv);

    sockaddr_in broadcast_addr;
    memset(&broadcast_addr, 0, sizeof(broadcast_addr));
    broadcast_addr.sin_family = AF_INET;
    broadcast_addr.sin_port = htons(DOIP_PORT);
    broadcast_addr.sin_addr.s_addr = inet_addr("255.255.255.255");

    DoIPHeader req_header;
    fill_header(&req_header, PAYLOAD_TYPE_VEHICLE_IDENT_REQ, 0);
    
    sendto(udp_sock, &req_header, sizeof(req_header), 0, (struct sockaddr*)&broadcast_addr, sizeof(broadcast_addr));

    std::string target_ip;
    uint16_t vehicle_logical_addr = 0;

    char buffer[1024];
    sockaddr_in from_addr;
    socklen_t from_len = sizeof(from_addr);

    ssize_t n = recvfrom(udp_sock, buffer, sizeof(buffer), 0, (struct sockaddr*)&from_addr, &from_len);
    if (n >= (ssize_t)sizeof(DoIPHeader)) {
        DoIPHeader* header = (DoIPHeader*)buffer;
        parse_header_ntoh(header);

        if (header->payload_type == PAYLOAD_TYPE_VEHICLE_IDENT_RES) {
            target_ip = inet_ntoa(from_addr.sin_addr);
            std::cout << "[UDP] 收到响应 来自 " << target_ip << std::endl;
            
            // Payload 偏移
            uint8_t* payload = (uint8_t*)(buffer + sizeof(DoIPHeader));
            
            // VIN 17 bytes
            std::string vin((char*)payload, 17);
            std::cout << "       -> 发现车辆 VIN: " << vin << std::endl;

            // Logical Addr
            vehicle_logical_addr = (payload[17] << 8) | payload[18];
            std::cout << "       -> 逻辑地址: 0x" << std::hex << vehicle_logical_addr << std::dec << std::endl;
        }
    } else {
        std::cout << "[!] 未发现车辆，尝试直连 127.0.0.1" << std::endl;
        target_ip = "127.0.0.1";
        vehicle_logical_addr = 0x1000;
    }
    close(udp_sock);

    if (target_ip.empty()) return 1;

    // 2. 建立 TCP 连接
    std::cout << "\n[Step 2] 连接车辆 TCP " << target_ip << ":" << DOIP_PORT << "..." << std::endl;
    int tcp_sock = socket(AF_INET, SOCK_STREAM, 0);
    sockaddr_in serv_addr;
    memset(&serv_addr, 0, sizeof(serv_addr));
    serv_addr.sin_family = AF_INET;
    serv_addr.sin_port = htons(DOIP_PORT);
    serv_addr.sin_addr.s_addr = inet_addr(target_ip.c_str());

    if (connect(tcp_sock, (struct sockaddr*)&serv_addr, sizeof(serv_addr)) < 0) {
        perror("[!] 连接失败");
        return 1;
    }

    // 3. 路由激活
    std::cout << "\n[Step 3] 发送路由激活请求..." << std::endl;
    // Source Address (2) + Activation Type (1) + Reserved (4)
    std::vector<uint8_t> ra_payload(7);
    ra_payload[0] = (TESTER_LOGICAL_ADDR >> 8) & 0xFF;
    ra_payload[1] = TESTER_LOGICAL_ADDR & 0xFF;
    ra_payload[2] = 0x00; // Type
    // Reserved 0...

    DoIPHeader ra_header;
    fill_header(&ra_header, PAYLOAD_TYPE_ROUTING_ACTIVATION_REQ, ra_payload.size());

    send(tcp_sock, &ra_header, sizeof(ra_header), 0);
    send(tcp_sock, ra_payload.data(), ra_payload.size(), 0);

    // 接收激活响应
    DoIPHeader res_header;
    if (recv_fixed(tcp_sock, &res_header, sizeof(res_header))) {
        parse_header_ntoh(&res_header);
        std::vector<uint8_t> res_payload(res_header.payload_length);
        if (recv_fixed(tcp_sock, res_payload.data(), res_header.payload_length)) {
            if (res_header.payload_type == PAYLOAD_TYPE_ROUTING_ACTIVATION_RES) {
                // ... Source ... Target ... Code (index 4)
                uint8_t code = res_payload[4];
                if (code == 0x10) {
                    std::cout << "[TCP] 路由激活成功! (Code 0x10)" << std::endl;
                } else {
                    std::cout << "[!] 路由激活失败: 0x" << std::hex << (int)code << std::dec << std::endl;
                    close(tcp_sock);
                    return 1;
                }
            }
        }
    }

    // 4. 发送诊断请求
    std::cout << "\n[Step 4] 发送诊断请求 (ReadDataByIdentifier 0x22 F1 90)..." << std::endl;
    // Payload: Source (2) + Target (2) + UDS Data
    std::vector<uint8_t> diag_payload;
    diag_payload.push_back((TESTER_LOGICAL_ADDR >> 8) & 0xFF);
    diag_payload.push_back(TESTER_LOGICAL_ADDR & 0xFF);
    diag_payload.push_back((vehicle_logical_addr >> 8) & 0xFF);
    diag_payload.push_back(vehicle_logical_addr & 0xFF);
    // UDS: 22 F1 90
    diag_payload.push_back(0x22);
    diag_payload.push_back(0xF1);
    diag_payload.push_back(0x90);

    DoIPHeader diag_header_req;
    fill_header(&diag_header_req, PAYLOAD_TYPE_DIAGNOSTIC_MESSAGE, diag_payload.size());
    
    send(tcp_sock, &diag_header_req, sizeof(diag_header_req), 0);
    send(tcp_sock, diag_payload.data(), diag_payload.size(), 0);

    // 5. 接收响应
    std::cout << "\n[Step 5] 等待响应..." << std::endl;
    while (true) {
        DoIPHeader h;
        if (!recv_fixed(tcp_sock, &h, sizeof(h))) break;
        parse_header_ntoh(&h);

        std::vector<uint8_t> p(h.payload_length);
        if (h.payload_length > 0) {
            if (!recv_fixed(tcp_sock, p.data(), h.payload_length)) break;
        }

        if (h.payload_type == PAYLOAD_TYPE_DIAGNOSTIC_POS_ACK) {
            std::cout << "[TCP] 收到 DoIP ACK (Positive)" << std::endl;
        } else if (h.payload_type == PAYLOAD_TYPE_DIAGNOSTIC_MESSAGE) {
            uint16_t sa = (p[0] << 8) | p[1];
            uint16_t ta = (p[2] << 8) | p[3];
            std::cout << "[TCP] 收到诊断响应 SA=0x" << std::hex << sa << " TA=0x" << ta << std::dec << std::endl;
            
            std::cout << "       -> UDS Data: ";
            for (size_t i = 4; i < p.size(); i++) {
                printf("%02X ", p[i]);
            }
            std::cout << std::endl;
            break; // 演示结束
        }
    }

    std::cout << "\n=== 演示结束 ===" << std::endl;
    close(tcp_sock);
    return 0;
}
